File: core/src/test/scala/unit/kafka/api/ApiUtilsTest.scala
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 * 
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package kafka.api

import org.junit._
import org.scalatest.junit.JUnitSuite
import org.junit.Assert._
import scala.util.Random
import java.nio.ByteBuffer
import kafka.common.KafkaException
import kafka.utils.TestUtils

object ApiUtilsTest {
  val rnd: Random = new Random()
}

class ApiUtilsTest extends JUnitSuite {

  @Test
  def testShortStringNonASCII() {
    // Random-length strings
    for(i <- 0 to 100) {
      // Since we're using UTF-8 encoding, each encoded byte will be one to four bytes long 
      val s: String = ApiUtilsTest.rnd.nextString(math.abs(ApiUtilsTest.rnd.nextInt()) % (Short.MaxValue / 4))  
      val bb: ByteBuffer = ByteBuffer.allocate(ApiUtils.shortStringLength(s))
      ApiUtils.writeShortString(bb, s)
      bb.rewind()
      assertEquals(s, ApiUtils.readShortString(bb))
    }
  }

  @Test
  def testShortStringASCII() {
    // Random-length strings
    for(i <- 0 to 100) {
      val s: String = TestUtils.randomString(math.abs(ApiUtilsTest.rnd.nextInt()) % Short.MaxValue)  
      val bb: ByteBuffer = ByteBuffer.allocate(ApiUtils.shortStringLength(s))
      ApiUtils.writeShortString(bb, s)
      bb.rewind()
      assertEquals(s, ApiUtils.readShortString(bb))
    }

    // Max size string
    val s1: String = TestUtils.randomString(Short.MaxValue)
    val bb: ByteBuffer = ByteBuffer.allocate(ApiUtils.shortStringLength(s1))
    ApiUtils.writeShortString(bb, s1)
    bb.rewind()
    assertEquals(s1, ApiUtils.readShortString(bb))

    // One byte too big
    val s2: String = TestUtils.randomString(Short.MaxValue + 1) 
    try {
      ApiUtils.shortStringLength(s2)
      fail
    } catch {
      case e: KafkaException => {
        // ok
      }
    }
    try {
      ApiUtils.writeShortString(bb, s2)
      fail
    } catch {
      case e: KafkaException => {
        // ok
      }
    }
  }
}

File: streams/src/test/java/org/apache/kafka/streams/internals/ApiUtilsTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.kafka.streams.internals;

import org.junit.jupiter.api.Test;

import java.time.Duration;
import java.time.Instant;

import static org.apache.kafka.streams.internals.ApiUtils.prepareMillisCheckFailMsgPrefix;
import static org.apache.kafka.streams.internals.ApiUtils.validateMillisecondDuration;
import static org.apache.kafka.streams.internals.ApiUtils.validateMillisecondInstant;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class ApiUtilsTest {
    // This is the maximum limit that Duration accepts but fails when it converts to milliseconds.
    private static final long MAX_ACCEPTABLE_DAYS_FOR_DURATION = 106751991167300L;
    // This is the maximum limit that Duration accepts and converts to milliseconds with out fail.
    private static final long MAX_ACCEPTABLE_DAYS_FOR_DURATION_TO_MILLIS = 106751991167L;

    @Test
    public void shouldThrowIllegalArgumentExceptionForNullDuration() {
        final String nullDurationPrefix = prepareMillisCheckFailMsgPrefix(null, "nullDuration");

        final IllegalArgumentException e = assertThrows(
            IllegalArgumentException.class,
            () -> validateMillisecondDuration(null, nullDurationPrefix)
        );
        assertTrue(e.getMessage().contains(nullDurationPrefix));
    }

    @Test
    public void shouldThrowArithmeticExceptionForMaxDuration() {
        final Duration maxDurationInDays = Duration.ofDays(MAX_ACCEPTABLE_DAYS_FOR_DURATION);
        final String maxDurationPrefix = prepareMillisCheckFailMsgPrefix(maxDurationInDays, "maxDuration");

        final IllegalArgumentException e = assertThrows(
            IllegalArgumentException.class,
            () -> validateMillisecondDuration(maxDurationInDays, maxDurationPrefix)
        );
        assertTrue(e.getMessage().contains(maxDurationPrefix));
    }

    @Test
    public void shouldThrowIllegalArgumentExceptionForNullInstant() {
        final String nullInstantPrefix = prepareMillisCheckFailMsgPrefix(null, "nullInstant");

        final IllegalArgumentException e = assertThrows(
            IllegalArgumentException.class,
            () -> validateMillisecondInstant(null, nullInstantPrefix)
        );
        assertTrue(e.getMessage().contains(nullInstantPrefix));
    }

    @Test
    public void shouldThrowArithmeticExceptionForMaxInstant() {
        final String maxInstantPrefix = prepareMillisCheckFailMsgPrefix(Instant.MAX, "maxInstant");

        final IllegalArgumentException e = assertThrows(
            IllegalArgumentException.class,
            () -> validateMillisecondInstant(Instant.MAX, maxInstantPrefix)
        );
        assertTrue(e.getMessage().contains(maxInstantPrefix));
    }

    @Test
    public void shouldReturnMillisecondsOnValidDuration() {
        final Duration sampleDuration = Duration.ofDays(MAX_ACCEPTABLE_DAYS_FOR_DURATION_TO_MILLIS);

        assertEquals(sampleDuration.toMillis(), validateMillisecondDuration(sampleDuration, "sampleDuration"));
    }

    @Test
    public void shouldReturnMillisecondsOnValidInstant() {
        final Instant sampleInstant = Instant.now();

        assertEquals(sampleInstant.toEpochMilli(), validateMillisecondInstant(sampleInstant, "sampleInstant"));
    }

    @Test
    public void shouldContainsNameAndValueInFailMsgPrefix() {
        final String failMsgPrefix = prepareMillisCheckFailMsgPrefix("someValue", "variableName");

        assertTrue(failMsgPrefix.contains("variableName"));
        assertTrue(failMsgPrefix.contains("someValue"));
    }
}
